core/server/common/src/main/java/alluxio/master/journal/raft/transport/CopycatGrpcConnection.java

```diff
@@ -0,0 +1,552 @@
+  private final Scheduled mTimeoutScheduler;
+
+  /** Used to synchronize during connection shut down. */
+  private final ReadWriteLock mStateLock;
+
+  /**
+   * Creates a connection object.
+   *
+   * Note: {@link #setTargetObserver} should be called explicitly before using the connection.
+   *
+   * @param connectionOwner owner of connection
+   * @param context copycat thread context
+   * @param requestTimeoutMs timeout in milliseconds for requests
+   */
+  public CopycatGrpcConnection(ConnectionOwner connectionOwner, ThreadContext context,
+      long requestTimeoutMs) {
+    mConnectionOwner = connectionOwner;
+    mContext = context;
+    mRequestTimeoutMs = requestTimeoutMs;
+
+    mStateLock = new ReentrantReadWriteLock();
+    mClosed = false;
+    mStreamCompleted = false;
+    mRequestCounter = new AtomicLong(0);
+    // Initialize timeout scheduler.
+    mTimeoutScheduler = context.schedule(Duration.ofMillis(mRequestTimeoutMs),
+        Duration.ofMillis(mRequestTimeoutMs / 2), this::timeoutPendingRequests);
+    // Initialize listeners.
+    mExceptionListeners = new Listeners<>();
+    mCloseListeners = new Listeners<>();
+    // Initialize handler and response maps.
+    mHandlers = new ConcurrentHashMap<>();
+    mResponseFutures = new ConcurrentSkipListMap<>();
+  }
+
+  /**
+   * Sets the stream observer for target.
+   *
+   * @param targetObserver target's stream observer
+   */
+  public void setTargetObserver(StreamObserver<CopycatMessage> targetObserver) {
+    mTargetObserver = targetObserver;
+  }
+
+  @Override
+  public CompletableFuture<Void> send(Object request) {
+    return sendAndReceiveInternal(request, true);
+  }
+
+  @Override
+  public <T, U> CompletableFuture<U> sendAndReceive(T request) {
+    return sendAndReceiveInternal(request, false);
+  }
+
+  /**
+   * Send the request to target. If "fireAndForget" then returned future will be complete.
+   *
+   * @param request request to send
+   * @param fireAndForget whether to not wait for response
+   * @param <T> Request type
+   * @param <U> Response type
+   * @return future for result
+   */
+  private <T, U> CompletableFuture<U> sendAndReceiveInternal(T request, boolean fireAndForget) {
+    try (LockResource lock = new LockResource(mStateLock.readLock())) {
+      Assert.notNull(request, "request");
+
+      // Create a contextual future for the request.
+      CopycatGrpcConnection.ContextualFuture<U> future =
+              new CopycatGrpcConnection.ContextualFuture<>(System.currentTimeMillis(),
+                      ThreadContext.currentContextOrThrow());
+
+      // Don't allow request if connection is closed.
+      if (mClosed) {
+        future.completeExceptionally(new IllegalStateException("Connection closed"));
+        return future;
+      }
+
+      // Get a new request Id.
+      long requestId = mRequestCounter.incrementAndGet();
+      // Register request future.
+      mResponseFutures.put(requestId, future);
```

Comment:
Do we need to do this if `fireAndForget` is true?

Reply:
That's for when `mTargetObserver.onNext()` takes longer than the timeout.